EB-3015 EB-4941 updates 1.0.3

nivhsay opened the pull request.

nivhsay committed:
updates

nivhsay commented:
@CodiumAI-Agent /review

QodoAI-Agent commented:
## PR Review 🔍

<table>
<tr>
<tr><td> ⏱️&nbsp;<strong>Estimated&nbsp;effort&nbsp;to&nbsp;review [1-5]</strong></td><td>

4, because the PR involves a significant amount of changes across multiple files including PHP classes and XML configuration. The changes impact core functionalities such as cart management and plugin lifecycle events, which require careful review to ensure compatibility and functionality within the Shopware platform.


</td></tr>
<tr><td> 🧪&nbsp;<strong>Relevant tests</strong></td><td>

No


</td></tr>
<tr><td rowspan=2> ⚡&nbsp;<strong>Possible issues</strong></td>
<td>

<strong>Possible Bug:</strong> The method `createSpriiCart` sets `lineItem->setStackable(false)` after adding the line item to the collection. This could lead to unexpected behavior as the stackable property should be determined before adding to the collection.</td></tr>
<tr>
<td>

<strong>Performance Concern:</strong> The method `getShopwareTaxIds` executes a database query inside a loop which can lead to performance issues if the number of products is large. Consider refactoring to reduce the number of queries.</td></tr>
<tr><td> 🔒&nbsp;<strong>Security concerns</strong></td><td>

No

</td></tr>
</table>


<details><summary> <strong>Code feedback:</strong></summary>

<hr><table><tr><td>relevant file</td><td>src/Services/SpriiCartService.php
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Consider setting the `stackable` property of line items before adding them to the collection to avoid potential issues with cart calculations. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/elisaio/elisa-shopware6/pull/2/files#diff-b9f85be1106d422c7a325c15b598d3117be38fa73e25687e8f3f34473ddd0eddR117'>$lineItem->setStackable(false);</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>src/Services/SpriiCartService.php
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Optimize the `getShopwareTaxIds` method by collecting all product IDs first and then executing a single query to fetch all tax IDs, instead of querying within a loop. This change will improve performance, especially with a large number of products. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/elisaio/elisa-shopware6/pull/2/files#diff-b9f85be1106d422c7a325c15b598d3117be38fa73e25687e8f3f34473ddd0eddR150'>return $this->connection->executeQuery(</a></td></tr></table><hr>

</details>

nivhsay committed:
Fix data loss issues in plugin lifecycle management

- Prevent data loss during plugin updates by using safe upsert operations
- Respect user data retention choice during uninstall via keepUserData()
- Replace dangerous removeExistingCustomFields with safe alternatives
- Preserve Sprii cart reference IDs and export status across plugin lifecycle
- Add explicit custom field cleanup for thorough uninstall when requested
- Remove unused helper methods (customFieldSetExists, legacy removeExistingCustomFields)

This ensures that historical Sprii order data is never accidentally deleted
during plugin updates while still allowing clean removal when explicitly chosen.

nivhsay committed:
shopware 6.6 updates

nivhsay committed:
6.7

QodoAI-Agent commented:
## PR Reviewer Guide 🔍

Here are some key observations to aid the review process:

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 3 🔵🔵🔵⚪⚪</td></tr>
<tr><td>🧪&nbsp;<strong>No relevant tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Recommended focus areas for review</strong><br><br>

<details><summary><a href='https://github.com/elisaio/elisa-shopware6/pull/2/files#diff-b9f85be1106d422c7a325c15b598d3117be38fa73e25687e8f3f34473ddd0eddR66-R85'><strong>Possible Issue</strong></a>

Switching to ProductLineItemFactory with an options array changes behavior versus the previous factory usage. Validate that the created line item type, referencedId, and quantity are correct for variants and that pricing/tax calculation still works as intended across Shopware 6.6/6.7.
</summary>

```php
$cart->setExtensions($cartExtensions);

$this->setAllowSpriiPriceOnCart($context);
$taxIds = $this->getShopwareTaxIds($spriiProducts);

// Loop through products from Sprii to create valid Shopware line items
foreach ($spriiProducts as $spriiProduct) {
    $productId = $spriiProduct['child_sku'] ?? $spriiProduct['sku'];
    $lineItem = $this->lineItemFactory->create([
        'id' => $productId,
        'referencedId' => $productId,
        'quantity' => $spriiProduct['qty']
    ], $context);

    // Set necessary fields on Shopware line item
    $lineItem->setStackable(true);
    $lineItem->setPayloadValue(SpriiLiveShoppingIntegration::SPRII_CART_LINEITEM, true);
    $lineItem->setPayloadValue(SpriiLiveShoppingIntegration::SPRII_SET_PRICE, false);

    if (isset($spriiProduct['price'])) {
```

</details>

<details><summary><a href='https://github.com/elisaio/elisa-shopware6/pull/2/files#diff-07d4160dab6ef49ce141932e7a68aa07b5ed2dea6b811dc6c48abbf1bda5d7e5R170-R201'><strong>Data Deletion</strong></a>

The uninstall logic deletes custom field sets and fields. Ensure the criteria and IDs used for deletion match actual records; current loops iterate over search results’ getElements keys, which may not be entity IDs.
</summary>

```php
public function removeCustomFieldSetAndData(Context $context): void
{
    $customFieldSetRepository = $this->container->get('custom_field_set.repository');
    $customFieldRepository = $this->container->get('custom_field.repository');

    // First, remove the custom field set
    $criteria = new Criteria();
    $criteria->addFilter(
        new EqualsFilter('custom_field_set.name', self::SPRII_ORDER_FIELDS)
    );

    $customFieldSet = $customFieldSetRepository->search($criteria, $context);

    foreach ($customFieldSet->getEntities()->getElements() as $key => $customField) {
        $customFieldSetRepository->delete([["id" => $key]], $context);
    }

    // Then, explicitly remove any orphaned custom fields
    $criteria = new Criteria();
    $criteria->addFilter(
        new EqualsAnyFilter(
            'custom_field.name',
            [self::SPRII_CART_REFERENCE_ID, self::SPRII_CART_EXPORTED]
        )
    );

    $customFields = $customFieldRepository->search($criteria, $context);

    foreach ($customFields->getEntities()->getElements() as $key => $customField) {
        $customFieldRepository->delete([["id" => $key]], $context);
    }
}
```

</details>

<details><summary><a href='https://github.com/elisaio/elisa-shopware6/pull/2/files#diff-d8f56ab452dfe178bc0d05b6dfa251a3dfa63454e2fa72c3afeae932bd80fbabR45-R56'><strong>Timestamp Logic</strong></a>

Building the order payload uses strtotime on only the day part of createdAt, which likely produces an incorrect epoch timestamp. Confirm Sprii expects full timestamp and timezone handling.
</summary>

```php
protected function buildSpriiOrder(OrderEntity $order): array
{
    $spriiCartId = $order->getCustomFields()[SpriiLiveShoppingIntegration::SPRII_CART_REFERENCE_ID];

    return [
        'orders' => [
            'id' => $order->getOrderNumber(),
            'sprii_reference' => $spriiCartId,
            'time' => strtotime($order->getCreatedAt()->format('d')),
            'products' => $this->buildSpriiProducts($order->getLineItems())
        ]
    ];
```

</details>

</td></tr>
</table>
